### PR TITLE
A complete rewrite of this library

### PR DESCRIPTION
**Pull request by BurningWitness**
I had trouble drawing bitmap characters with `freetype-0.1.2`, so I decided to update this library a tiny bit.

This PR:
- Adds every single binding one can find in the [API reference](https://www.freetype.org/freetype2/docs/reference/index.html), basing module names on the API reference chapters; `FreeType.Core.Mac` and it's `Internal` are commented out, as I could not find a way to link `ftmac.h` properly under OS X.
- Moves every enumeration to patterns, therefore ramping the minimum base version to ~~`7.8.1`~~ `8.2.2`;
- Wraps a lot of functions for user convenience, some in a rather weird fashion (e.g. in `FreeType.Support.Outline` and `FreeType.Support.Bitmap` initialization functions return `malloc`'ed pointers, as all other functions in those modules work specifically with pointers). Every modified function has a raw version in that module's `Internal` module;
- Adds 398 lens for every single datatype;
- Makes functions throw `FtError` exception so you never have to check the return values;
- Checks for whether the names of dynamic functions are bound properly in a separate test;
- Rewrites the test executable to a smaller nicer one;
- Definitely breaks [six packages](http://packdeps.haskellers.com/reverse/freetype2), none of which are on Stackage;
- Has everything Haddock-documented in a way that matches the API reference;
- Updates `.travis.yml`. Unfortunately `stack` compiled under `GHC 8.2.2` throws an `Access violation` error when compiling Template Haskell code under 32-bit Windows, so that part's commented out;

Mild inconveniences:
- All of the patterns are non-polymorphic, so this package is most probably `fromIntegral`s galore (as with `FT_Pixel_Mode` that does not match the `bitmap.pixel_mode`);

I am gravely sorry for the insurmountable size of this PR. Seeing how low the usage of this library has been over the years, I would not blame you if you decided against it.

### CODE REVIEW

File: FreeType/Exception/Internal.hsc [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+-- | An internal function for throwing 'FtError's
+ftError :: Name -> IO FT_Error -> IO ()
+ftError name action = do
+  code <- action
+  unless (code == 0)
+    . throwIO $ FtError (nameBase name) code
+
+
+
+class AutoError a b where
```
Comment (Mokosha) [inline FreeType/Exception/Internal.hsc:27]: This seems unnecessary. You're saving very little by calling `autoError` instead of `ftError`. It seems like it would be clearer to leave it with the latter in all the call sites (notice in `AutoAllocaError` you used `ftError` when you could have used `autoError`).

File: FreeType/Exception/Internal.hsc [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+      ftError name $ f ptr
+      peek ptr
+
+instance Storable b
+      => AutoAllocaError (a -> Ptr b -> IO FT_Error)
+                         (a -> IO b) where
+  autoAllocaError name f = \a ->
+    alloca $ \ptr -> do
+      ftError name $ f a ptr
+      peek ptr
```
Comment (Mokosha) [inline FreeType/Exception/Internal.hsc:67]: This, and every subsequent instance can be written as a special case of the first:

`autoAllocaError name f = autoAllocaError name . f`

... making this typeclass just as redundant as `AutoError` above.

File: FreeType/Error/Values.hsc [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+      , FT_Err_Missing_Fontboundingbox_Field
+      , FT_Err_Missing_Chars_Field
+      , FT_Err_Missing_Startchar_Field
+      , FT_Err_Missing_Encoding_Field
+      , FT_Err_Missing_Bbx_Field
+      , FT_Err_Bbx_Too_Big
+      , FT_Err_Corrupted_Font_Header
+      , FT_Err_Corrupted_Font_Glyphs
+     :: FT_Error
+pattern FT_Err_Ok                            = #const FT_Err_Ok
```
Comment (Mokosha) [inline FreeType/Error/Values.hsc:205]: Why are these pattern synonyms and not something like `Num a => a`, or even just of type `FT_Int` ?

File: FreeType.hs [hunk 4]
```diff
@@ -0,0 +1,97 @@
+{-# OPTIONS_GHC -fno-warn-dodgy-exports #-}
+{-# OPTIONS_GHC -fno-warn-unused-imports #-}
```
Comment (Mokosha) [inline FreeType.hs]: These seem like legitimate warnings. Why are they turned off?

File: FreeType/Cache.hsc [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+{- | Please refer to the
+     [Cache Sub-System > Cache Sub-System](https://www.freetype.org/freetype2/docs/reference/ft2-cache_subsystem.html)
+     chapter of the reference.
+
+     Internal: "FreeType.Cache.Internal".
+ -}
+
+module FreeType.Cache
+  ( -- ** FTC_Manager
```
Comment (Mokosha) [inline FreeType/Cache.hsc:11]: These kinds of haddock headings are pretty much useless if they only copy the type name. Perhaps sort them in a way that makes sense or remove them?

File: FreeType/Cache.hsc [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+  -> (FTC_Manager -> IO a)
+  -> IO a
+ftc_Manager_With lib faces sizes bytes req data_ =
+  bracket (ftc_Manager_New lib faces sizes bytes req data_) ftc_Manager_Done
+
+
+
+foreign import ccall "FTC_Manager_Reset"
+  ftc_Manager_Reset
+    :: FTC_Manager -- ^ manager
```
Comment (Mokosha) [inline FreeType/Cache.hsc:117]: Similarly, a comment like this: `-- ^ manager`, is more or less not useful. I'd remove it to avoid clutter in the haddock (here and everywhere).

File: FreeType/Lens.hs [hunk 7]
```diff
@@ -0,0 +1,253 @@
+{-# LANGUAGE FlexibleInstances #-}
```
Comment (Mokosha) [inline FreeType/Lens.hs]: I feel like this file can be split into its own library that depends on lens and/or microlens. I don't use lens in any of my projects and I don't particularly want the unnecessary dependency.

File: FreeType/Format/Multiple.hsc [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+setCoords name f face coords =
+  withArray coords $ ftError name . f face (fromIntegral $ length coords)
+
+
+
+getCoords
+  :: Storable a
+  => Name
+  -> (FT_Face -> FT_UInt -> Ptr a -> IO FT_Error)
+  -> (FT_Face -> FT_UInt -> IO [a])
```
Comment (Mokosha) [inline FreeType/Format/Multiple.hsc:114]: It's a shame that we're going from unboxed `Ptr a` to boxed `[a]`. This seems like a decision the caller should be allowed to make. What do you think about keeping this interface as `Ptr a`: some callers might want to use `allocaArray` as you have, but others would want (e.g.) the `StorableArray` interface?

File: FreeType/Support/Advance.hsc [hunk 9]
```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE TemplateHaskell #-}
```
Comment (Mokosha) [inline FreeType/Support/Advance.hsc]: Do you need TemplateHaskell here (this applies to some other modules on second glance, too)?

**Review by Mokosha — COMMENTED** [review]
Hi!

Thank you so much for taking the time to update this library. This looks like it was a lot of work and it's definitely appreciated!!

That being said, this was a rather large change, so I gave it a once-over on review comments. There are a few high-level things that I think would make this PR a lot better:

1) Please remove all of the files from freetype2 that we don't need to build the library. In particular, the `docs` and `builds` folders probably don't need to be there.
2) Please document what exactly breaks from the old version. This will help people that want to convert (if ever).
3) Please try and keep all of the dependencies as minimal as possible. We shouldn't need lenses and template haskell for bindings to a C library. 
4) In your PR description (which was nice and detailed, thanks!) you mentioned GHC 8.2 -- Please also test it on newer versions of GHC, ideally at least 8.8.

In the future, it might be better to open an issue first to discuss some of the changes you're planning to make. Overall, though, I think the work here is pretty good, so the comments are mostly just clarifications and nitpicks.

I will be squashing before merging in order to reduce the size of the changes included in the main branch.

File: FreeType/Support/Advance.hsc (hunk 9 above)
Comment (BurningWitness) [reply FreeType/Support/Advance.hsc]: Pretty much all modules that throw `ftError`s require `TemplateHaskell`, because I pass the `Name` of the function instead of writing it as a `String` (so that I don't mistype a function name).

The `template-haskell` dependency is required by `microlens-th` already, so the only negative is a slightly longer compilation time in my eyes. However if the `microlens-th` is to be moved into a separate `freetype2-lens` package, then I guess I can just regex them into `String`s.

File: FreeType/Cache.hsc (hunk 5 above)
Comment (BurningWitness) [reply FreeType/Cache.hsc:11]: The module structure of this package, headers as well as the function annotations are nods to the [API reference](https://www.freetype.org/freetype2/docs/reference/ft2-cache_subsystem.html). It's definitely not the most `Haddock`y way of structuring things, but imo this one is not supposed to be viewed without the API reference open in another tab.

File: FreeType/Cache.hsc (hunk 6 above)
Comment (BurningWitness) [reply FreeType/Cache.hsc:117]: Certain functions this library exports do not match the definitions of the API reference argument-order-wise, so these are mostly for that. I do however agree that 95% of them are absolutely useless, but I don't think it clutters the Haddock that much, so I'm split on it.

File: FreeType/Error/Values.hsc (hunk 3 above)
Comment (BurningWitness) [reply FreeType/Error/Values.hsc:205]: From the get go I decided to use patterns simply because all of those names have `FT` capitalized in the FreeType library and `gl` already does export every value of this kind as patterns.

However the part where you make everything `(Eq a, Num a) => a` is a bit wacky here, because certain definitions (e.g. [FT_Glyph_Format](https://www.freetype.org/freetype2/docs/reference/ft2-basic_types.html#ft_glyph_format)) actually get used as types later on (e.g. in [FT_GlyphSlotRec](https://www.freetype.org/freetype2/docs/reference/ft2-base_interface.html#ft_glyphslotrec)), while some do not.

So I guess I should leave the types be and set every pattern to `(Eq a, Num a) => a` nonetheless.

File: FreeType.hs (hunk 4 above)
Comment (BurningWitness) [reply FreeType.hs]: This stems from the fact that I never managed to make `FreeType.Core.Mac.Internal` link properly on OS X, so I decided to leave that module reexported in `FreeType` (and `FreeType.Internal`), despite both `FreeType.Core.Mac` and `FreeType.Core.Mac.Internal` exporting absolutely nothing.

I will remove those reexports.

File: FreeType/Exception/Internal.hsc (hunk 1 above)
Comment (BurningWitness) [reply FreeType/Exception/Internal.hsc:27]: `AutoError` and `AutoAllocaError` are merely convenience functions for nice clean copy-pasting of functions so one never has to retype up to seven arguments, e.g.
```
ft_Get_Track_Kerning :: FT_Face -> FT_Fixed -> FT_Int -> IO FT_Fixed
ft_Get_Track_Kerning =
  autoAllocaError 'ft_Get_Track_Kerning ft_Get_Track_Kerning'
```

It is absolutely true that you can just use `ftError` in such cases, however I think this is much more clean.

File: FreeType/Format/Multiple.hsc (hunk 8 above)
Comment (BurningWitness) [reply FreeType/Format/Multiple.hsc:114]: I do not think this is an issue, as if anyone requires tight loops for efficiency can just rewrite the functions by using definitions in `FreeType.Format.Multiple.Internal`.

File: FreeType/Lens.hs (hunk 7 above)
Comment (BurningWitness) [reply FreeType/Lens.hs]: This is a rather tough question, because on one hand the library has **a lot** of overlapping field names, so you'd pretty much want to use `[micro]lens`, however if you actually do, you cannot `peek` with them, allowing for some hilarious one-liners like
```
\face -> do
   btmp <- view bitmap <<$>> peek . view glyph =<< peek face
```

Though now that I think of it I should just regex all the `template-haskell` `Name`s in modules, you're right.

**Comment by BurningWitness** [comment]
Extra things:

2. This update breaks quite literally everything, as the previous iteration resided under `Graphics.Rendering.FreeType.Internal`, datatypes had field overlaps and were therefore spread across a dozen `Internal.*` modules (so you had to import about seven modules to do anything) and instead of full-on Storable instances you had `:: a -> Ptr b` functions. I genuinely tried to base old modules on the new ones, but that required a lot of `Integral` conversions and some separate `hsc2hs` usage, so I gave up on it. Therefore I have no idea on how to document the migration guide other than "rewrite from scratch", sadly.

4. I updated the Travis CI, it's compiles on `8.2.2`, `8.4.4`, `8.6.5` and `8.8.3` across Linux and OS X. The only problems I had there are `cabal 8.8.3`, the CI version of which uses a bad version of `th-abstractions` and the Windows build that pretty much always fails due to an access violation due to `GHC 8.2.2` bugs on Win32.

File: FreeType/Exception/Internal.hsc (hunk 1 above)
Comment (Mokosha) [reply FreeType/Exception/Internal.hsc:27]: OK -- It wasn't clear to me that the maintenance burden of the rest of the library was that high. Thanks for the clarification.

File: FreeType/Error/Values.hsc (hunk 3 above)
Comment (Mokosha) [reply FreeType/Error/Values.hsc:205]: OK -- I would use `(Eq a, Num a) => a` where it makes sense for consistency with other packages and keep it as is otherwise. For things like `FT_Error` I would expect a similar type enum like `FT_Glyph_Format`. Looking over the files again, you seem to have used good judgement where applicable, so I trust you'll make the appropriate changes. :)

File: FreeType/Cache.hsc (hunk 5 above)
Comment (Mokosha) [reply FreeType/Cache.hsc:11]: Please add that note to the README then (prior to all of the stuff pulled from the original freetype2 library).

File: FreeType/Cache.hsc (hunk 6 above)
Comment (Mokosha) [reply FreeType/Cache.hsc:117]: These make a lot more sense if you're expected to be looking at the freetype docs at the same time. Resolving it per my previous note.

File: FreeType/Format/Multiple.hsc (hunk 8 above)
Comment (Mokosha) [reply FreeType/Format/Multiple.hsc:114]: It's weird to me to export `Internal` modules -- seems like a contradiction, no?

**Comment by Mokosha** [comment]
FWIW, I think we only need to keep the most recent version of GHC (8.8.3), which should be fixed once you remove `template-haskell` and the nightly versions of Stack and Cabal (which can fail).

Expecting users to upgrade to newer GHC versions is not unreasonable, imo.

**Comment by BurningWitness** [comment]
I don't really think there is a point of removing 8.2.2 compatibility though, as it's already there. Also now all of Linux and OS X tests complete, nightlies are always allowed to fail and will be run once in a decade either way.

File: FreeType/Cache.hsc (hunk 5 above)
Comment (BurningWitness) [reply FreeType/Cache.hsc:11]: I already did and I removed that tediously long description for a "Refer to README"

File: FreeType/Format/Multiple.hsc (hunk 8 above)
Comment (BurningWitness) [reply FreeType/Format/Multiple.hsc:114]: They're mostly there to not pile stuff together (`Internal` modules storing foreign imports for functions I wrapped and deprecations), the user is supposed to use `FreeType` and _if they wish_ `FreeType.Internal`. Well and there's also `FreeType.Exception` for `FtError`.

File: freetype2.cabal [hunk 10]
```diff
@@ -1,88 +1,10 @@
...
-                     .
-                     FreeType 2 provides its own caching subsystem
-                     since release 2.0.1. It can be used to cache
-                     either face instances or glyph images
-                     efficiently.
+                     Please check the README for information on how these
+                     bindings operate.
 License:             BSD3
 License-file:        LICENSE
 Author:              Jason Dagit <dagitj@gmail.com>
```
Comment (Mokosha) [inline freetype2.cabal]: Please feel free to add yourself as an author here. :)

**Comment by BurningWitness** [comment]
So I moved the `FreeType.Lens` to a separate [`freetype-lens`](https://github.com/BurningWitness/freetype2-lens) package, however I do not know how the overlap of it relying on this package despite it not being on Stackage is to be resolved (so I assume the `stack.yaml` and `cabal.project` in `freetype-lens` will be removed somehow; and also I could not figure out how to wire up a `cabal.project` as I'm using stack).

Other than that it seems to be done, unless you have any other suggestions.

**Comment by Mokosha** [comment]
You should wait until this package is updated, then you should be free to upload freetype-lens. I'll let you know when I do that here.

File: test/Fields/Lens.hsc [hunk 11]
```diff
@@ -0,0 +1,52 @@
+module Fields.Lens where
```
Comment (Mokosha) [inline test/Fields/Lens.hsc:1]: Is this file still needed?

Comment (BurningWitness) [reply test/Fields/Lens.hsc:1]: It's a test for whether all fields are prefixed correctly (which also ensures lens can be generated by `freetype-lens`), not a library bit. Unless that somehow adds `template-haskell` as a library requirement for users, but I don't think it does.

Comment (Mokosha) [reply test/Fields/Lens.hsc:1]: I think depending on `template-haskell` in the test-suite is OK. I'm on the fence about whether or not it's needed though, since there shouldn't really be a connection between this library and `freetype-lens`, but given that it's not in the main library I'm inclined to keep it.

**Comment by Mokosha** [comment]
I tried building this on Windows just now. It failed because it couldn't find the `FT_Int64` and `FT_UInt64` types. After commenting them out on the Haskell side, everything built and I successfully ran the example program.

Don't know how you want to handle this (it seems like these types aren't used anywhere in the bindings and are only needed internally).

**Comment by BurningWitness** [comment]
Those types are indeed not used at all across the library so I think it's safe to comment them out under Windows. I do not know whether this issue manifests only on 32-bit versions of Windows, but I don't think it actually matters in this case, as these are merely type synonyms.

Scrap that, I'll just set it to
```
#ifndef mingw32_HOST_OS
type FT_Int64 = #type FT_Int64
#else
type FT_Int64 = Int64
#endif
```

I assume the freetype2 comment which states
```
A typedef for a 64bit [un]signed integer type.  The size depends on the
configuration.  Only defined if there is real 64bit support;
otherwise, it gets emulated with a structure (if necessary).
```
does not actually matter over FFI, since the final structure will take up eight bytes either way.

**Comment by Mokosha** [comment]
LGTM. Thanks for making the changes.

**Comment by Mokosha** [comment]
I've uploaded the [new version](https://hackage.haskell.org/package/freetype2) to hackage -- feel free to use it for `freetype2-lens`
